fe/fe-core/src/main/java/com/starrocks/planner/PlanNode.java

```diff
@@ -710,16 +712,132 @@ public boolean canPushDownRuntimeFilter() {
     public void checkRuntimeFilterOnNullValue(RuntimeFilterDescription description, Expr probeExpr) {
     }
 
-    public boolean pushDownRuntimeFilters(RuntimeFilterDescription description, Expr probeExpr) {
+    /**
+     * Return candidate slot exprs which is same to the expr, eg: tb1 join tb2 on tb1.a = tb2.b, when expr param is tb1.a,
+     * tb2.b is the candidate slot expr for tb1.a which has the same syntax for the query.
+     *
+     * @param expr: the slot expr that need to find its candidate slot exprs.
+     * @return List<Expr>: all the slot expr's candidate slot exprs.
+     */
+    public List<Expr> candidatesOfSlotExpr(Expr expr) {
+        List<Expr> candidateSlotExprs = Lists.newArrayList();
+        // NOTE: No need to check expr is slot or not here, each node should implement its `candidatesOfSlotExpr` itself.
+        if (!expr.isBoundByTupleIds(getTupleIds())) {
+            return candidateSlotExprs;
+        }
+        // only contain the input expr by default.
+        candidateSlotExprs.add(expr);
+        return candidateSlotExprs;
+    }
+
+    public List<List<Expr>> candidatesOfSlotExprs(List<Expr> exprs) {
+        List<List<Expr>> candidatesOfSlotExprs = Lists.newArrayList();
+        for (Expr expr: exprs) {
+            List<Expr> candidates = candidatesOfSlotExpr(expr);
+            if (candidates.isEmpty()) {
+                return Lists.newArrayList();
+            }
+            candidatesOfSlotExprs.add(candidates);
+        }
+        return permutaionsOfPartitionByExprs(candidatesOfSlotExprs);
+    }
+
+    public static List<List<Expr>> permutaionsOfPartitionByExprs(List<List<Expr>> partitionByExprs) {
```

Comment:
What's the goal of this method?

Reply:
This method is used for `partitionByExprs`: when each partitionByExpr has many candidates, use this method to generate all possible permutations that are used to pushed down.

eg, parttionByExprs: col1, col2
col1's candidates: col1_a, col1_b
col2's candidates: col2_a, col2_b

The possible permutations are 
- col1_a, col2_a
- col1_a, col2_a
- col1_b, col2_a
- col1_b, col2_b

Once one permuatation is satisfied, the loop will be broken.